sdk/python/lib/pulumi/errors.py

```diff
@@ -12,9 +12,41 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+from typing import List, Optional
+
 
 class RunError(Exception):
     """
     Can be used for terminating a program abruptly, but resulting in a clean exit rather than the usual
     verbose unhandled error logic which emits the source program text and complete stack trace.
     """
+
+
+class InputPropertyError(Exception):
+    def __init__(self, message: str, property_path: str, reason: str):
+        """
+        Can be used to indicate that the client has made a request with a bad input property.
+        """
+        self.message = message
+        self.property_path = property_path
+        self.reason = reason
+
+
+class PropertyError:
+    """
+    Represents an error in a property value.
+    """
+
+    def __init__(self, property_path: str, reason: str):
+        self.property_path = property_path
+        self.reason = reason
+
+
+class InputPropertiesError(Exception):
+    def __init__(self, message: str, errors: Optional[List[PropertyError]] = None):
```

Comment:
I wonder if there's a better way to pass this list of errors here.  Does someone with more python experience have any ideas?

Reply:
We could do what we do for our SDKs and think about using `TypedDicts`; there is also a fluent-style API option potentially:

TypedDicts
```python
def __init__(self, message: str, errors: Optional[List[PropertyError | PropertyErrorDict]] = None)

# Potentially all options depending on how deep we take the union
raise InputPropertiesError("foo", [{"property_path": "x", "reason": "y"}, ...])
raise InputPropertiesError("foo", [PropertyError(property_path="x", reason="y"), ...])
raise InputPropertiesError("foo", [dict(property_path="x", reason="y"), ...])
```

Fluent
```python
raise InputPropertiesError("foo").with_property(property_path="x", reason="y")...
```

There is likely the ability to offer both too if we think that's valuable.

Reply:
Ah I forgot about `TypeDicts`.  Imo that seems like the best option here, I'm not even sure we should allow the others, since that seems to be the style we're starting to follow, and seems pretty clean.

Reply:
typeddicts feels worthwhile, not sure I'd bother with fluent style to start.